src/random.cpp

```diff
@@ -107,36 +166,37 @@ static void InitHardwareRand() {}
 static void ReportHardwareRand() {}
 #endif
 
-static bool GetHardwareRand(unsigned char* ent32) noexcept {
+static bool GetHardwareRand64(unsigned char* ent8) noexcept {
 #if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
-    if (rdrand_supported) {
-        uint8_t ok;
-        // Not all assemblers support the rdrand instruction, write it in hex.
-#ifdef __i386__
-        for (int iter = 0; iter < 4; ++iter) {
-            uint32_t r1, r2;
-            __asm__ volatile (".byte 0x0f, 0xc7, 0xf0;" // rdrand %eax
-                              ".byte 0x0f, 0xc7, 0xf2;" // rdrand %edx
-                              "setc %2" :
-                              "=a"(r1), "=d"(r2), "=q"(ok) :: "cc");
-            if (!ok) return false;
-            WriteLE32(ent32 + 8 * iter, r1);
-            WriteLE32(ent32 + 8 * iter + 4, r2);
-        }
-#else
-        uint64_t r1, r2, r3, r4;
-        __asm__ volatile (".byte 0x48, 0x0f, 0xc7, 0xf0, " // rdrand %rax
-                                "0x48, 0x0f, 0xc7, 0xf3, " // rdrand %rbx
-                                "0x48, 0x0f, 0xc7, 0xf1, " // rdrand %rcx
-                                "0x48, 0x0f, 0xc7, 0xf2; " // rdrand %rdx
-                          "setc %4" :
-                          "=a"(r1), "=b"(r2), "=c"(r3), "=d"(r4), "=q"(ok) :: "cc");
-        if (!ok) return false;
-        WriteLE64(ent32, r1);
-        WriteLE64(ent32 + 8, r2);
-        WriteLE64(ent32 + 16, r3);
-        WriteLE64(ent32 + 24, r4);
+    if (g_rdrand_supported) {
+        uint64_t out = GetRdRand();
+        WriteLE64(ent8, out);
+        return true;
+    }
 #endif
+    return false;
+}
+
+static bool GetHardwareRand256(unsigned char* ent32) noexcept {
+#if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+    // When we want 256 bits of entropy, prefer RdSeed over RdRand, as it's
+    // guaranteed to produce independent randomness on every call. RdRand
+    // output is generated using an RNG seeded by the actual entropy, which does
+    // not guarantee (information theoretically) independent results.
+    if (g_rdseed_supported) {
+        uint64_t out1 = GetRdSeed(), out2 = GetRdSeed(), out3 = GetRdSeed(), out4 = GetRdSeed();
+        WriteLE64(ent32, out1);
+        WriteLE64(ent32 + 8, out2);
+        WriteLE64(ent32 + 16, out3);
+        WriteLE64(ent32 + 24, out4);
+        return true;
+    }
+    if (g_rdrand_supported) {
+        uint64_t out1 = GetRdRand(), out2 = GetRdRand(), out3 = GetRdRand(), out4 = GetRdRand();
+        WriteLE64(ent32, out1);
+        WriteLE64(ent32 + 8, out2);
+        WriteLE64(ent32 + 16, out3);
+        WriteLE64(ent32 + 24, out4);
         return true;
```

Comment:
Maybe we should ignore it for the less critical runtime use, but pedantically: 

"After invoking the RDRAND instruction, the caller must examine the carry flag (CF) to determine whether a random value was available at the time the RDRAND instruction was executed. As Table 3 shows, a value of 1 indicates that a random value was available and placed in the destination register provided in the invocation. A value of 0 indicates that a random value was not available. In current architectures the destination register will also be zeroed as a side effect of this condition." ... "It is recommended that applications attempt 10 retries in a tight loop in the unlikely event that the RDRAND instruction does not return a random number. This number is based on a binomial probability argument: given the design margins of the DRNG, the odds of ten failures in a row are astronomically small and would in fact be an indication of a larger CPU issue."

This one might also want to consider:

"There are two approaches to structuring RDRAND invocations such that DRBG reseeding can be guaranteed:

    Iteratively execute RDRAND beyond the DRBG upper bound by executing more than 1022 64-bit RDRANDs"